File: VB/ReplaceDefaultActionInListView.Module.Win/WriteMailController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Diagnostics
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.SystemModule

Namespace ReplaceDefaultActionInListView.Module
    Public Class WriteMailController
        Inherits ViewController
        Private writeMailAction As SimpleAction
        Public Sub New()
            TargetObjectType = GetType(AddressBookRecord)
            writeMailAction = New SimpleAction(Me, "WriteMail", PredefinedCategory.Edit)
            writeMailAction.ToolTip = "Write e-mail to the selected address book record"
            writeMailAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject
            writeMailAction.ImageName = "BO_Contact"
            AddHandler writeMailAction.Execute, AddressOf writeMailAction_Execute
        End Sub
        Private Sub writeMailAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            Dim record As AddressBookRecord = CType(e.CurrentObject, AddressBookRecord)
            Dim startInfo As String = String.Format("mailto:{0}?body=Hello, {1}!%0A%0A", record.Email, record.Name)
            Process.Start(startInfo)
        End Sub
        Protected Overrides Overloads Sub OnActivated()
            MyBase.OnActivated()
            Dim processCurrentObjectController As ListViewProcessCurrentObjectController = Frame.GetController(Of ListViewProcessCurrentObjectController)()
            If processCurrentObjectController IsNot Nothing Then
                AddHandler processCurrentObjectController.CustomProcessSelectedItem, AddressOf processCurrentObjectController_CustomProcessSelectedItem
            End If
        End Sub
        Private Sub processCurrentObjectController_CustomProcessSelectedItem(ByVal sender As Object, ByVal e As CustomProcessListViewSelectedItemEventArgs)
            e.Handled = True
            writeMailAction.DoExecute()
        End Sub
        Public Property DefaultListViewAction() As SimpleAction
            Get
                Return writeMailAction
            End Get
            Set(ByVal value As SimpleAction)
                writeMailAction = value
            End Set
        End Property
    End Class
End Namespace

File: CS/ReplaceDefaultActionInListView.Module.Win/WriteMailController.cs
using System;
using System.Diagnostics;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.SystemModule;

namespace ReplaceDefaultActionInListView.Module {
    public class WriteMailController : ViewController {
        private SimpleAction writeMailAction;
        public WriteMailController() {
            TargetObjectType = typeof(AddressBookRecord);
            writeMailAction = new SimpleAction(this, "WriteMail", PredefinedCategory.Edit);
            writeMailAction.ToolTip = "Write e-mail to the selected address book record";
            writeMailAction.SelectionDependencyType = SelectionDependencyType.RequireSingleObject;
            writeMailAction.ImageName = "BO_Contact";
            writeMailAction.Execute += writeMailAction_Execute;
        }
        void writeMailAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            AddressBookRecord record = (AddressBookRecord)e.CurrentObject;
            string startInfo =
                String.Format("mailto:{0}?body=Hello, {1}!%0A%0A", record.Email, record.Name);
            Process.Start(startInfo);
        }
        private ListViewProcessCurrentObjectController processCurrentObjectController;
        protected override void OnActivated() {
            base.OnActivated();
            processCurrentObjectController =
                Frame.GetController<ListViewProcessCurrentObjectController>();
            if (processCurrentObjectController != null) {
                processCurrentObjectController.CustomProcessSelectedItem +=
                    processCurrentObjectController_CustomProcessSelectedItem;
            }
        }
        private void processCurrentObjectController_CustomProcessSelectedItem(
            object sender, CustomProcessListViewSelectedItemEventArgs e) {
            e.Handled = true;
            writeMailAction.DoExecute();
        }
        public SimpleAction DefaultListViewAction {
            get { return writeMailAction; }
            set { writeMailAction = value; }
        }
        protected override void OnDeactivated() {
            if (processCurrentObjectController != null) {
                processCurrentObjectController.CustomProcessSelectedItem -= 
                    processCurrentObjectController_CustomProcessSelectedItem;
            }
            base.OnDeactivated();
        }
    }
}
